src/main/kotlin/com/wgslfuzz/core/Resolver.kt

```diff
@@ -510,71 +557,87 @@ private fun nodeIntroducesNewScope(
             parentNode !is ContinuingStatement
     )
 
+private fun <T : Type> T.asStoreType(): Type =
+    when (this) {
+        is Type.Reference -> this.storeType
+        else -> this
+    }
+
+private fun defaultAccessModeOf(addressSpace: AddressSpace): AccessMode =
+    when (addressSpace) {
+        AddressSpace.FUNCTION -> AccessMode.READ_WRITE
+        AddressSpace.PRIVATE -> AccessMode.READ_WRITE
+        AddressSpace.WORKGROUP -> AccessMode.READ_WRITE
+        AddressSpace.UNIFORM -> AccessMode.READ
+        AddressSpace.STORAGE -> AccessMode.READ
+        AddressSpace.HANDLE -> AccessMode.READ
+    }
+
 private fun resolveExpressionType(
     expression: Expression,
     resolverState: ResolverState,
 ): Type =
     when (expression) {
         is Expression.FunctionCall -> resolveTypeOfFunctionCallExpression(expression, resolverState)
-        is Expression.IndexLookup ->
+        is Expression.IndexLookup -> {
+            fun resolveDirectIndexType(type: Type) =
+                when (type) {
+                    is Type.Matrix -> Type.Vector(width = type.numRows, elementType = type.elementType)
+                    is Type.Vector -> type.elementType
+                    is Type.Array -> type.elementType
+                    else -> throw IllegalArgumentException("Index lookup attempted on unsuitable type $type")
+                }
+
             when (val targetType = resolverState.resolvedEnvironment.typeOf(expression.target)) {
-                is Type.Matrix ->
-                    Type.Vector(
-                        width = targetType.numRows,
-                        elementType = targetType.elementType,
-                    )
-                is Type.Vector -> targetType.elementType
-                is Type.Array -> targetType.elementType
                 is Type.Pointer -> {
-                    when (val pointeeType = targetType.pointeeType) {
-                        is Type.Vector -> pointeeType.elementType
-                        is Type.Array -> pointeeType.elementType
-                        is Type.Matrix -> Type.Vector(pointeeType.numRows, pointeeType.elementType)
-                        else -> TODO("Index lookup on pointer with pointee type ${targetType.pointeeType}")
-                    }
+                    val newPointeeType = resolveDirectIndexType(targetType.pointeeType)
+                    Type.Pointer(newPointeeType, targetType.addressSpace, targetType.accessMode)
                 }
-                else -> throw IllegalArgumentException("Index lookup attempted on unsuitable type $targetType")
+                is Type.Reference -> {
+                    val newStoreType = resolveDirectIndexType(targetType.storeType)
```

Comment:
Regarding my request for examples above in the pointer case, a quick example here would be good too - presumably it's fairly similar to the pointer case so can be briefer.